File: 2024/Day 1/src/A.kt
import java.io.File
import kotlin.math.abs

fun main() {
    // val input = File("../example input.txt")
    val input = File("../input.txt")

    val left = mutableListOf<Int>()
    val right = mutableListOf<Int>()

    input.forEachLine() { line ->
        // Splitting the two numbers on each line by whatever whitespace separates them
        val (leftValue, rightValue) = line.split("\\s+".toRegex()).map(String::toInt)
        left.add(leftValue)
        right.add(rightValue)
    }

    left.sort()
    right.sort()

    val delta = left.zip(right) {left, right -> abs(left - right)}

    println("Left: $left")
    println("Right: $right")
    println("Delta: $delta")
    println("Sum of deltas: ${delta.sum()}")
}

File: 2024/Day 1/src/bin/a.rs
use parse_display::FromStr;
use std::fs::read_to_string;

#[derive(FromStr, Debug)]
#[display("{left}   {right}")]
struct Line {
    left: u32,
    right: u32,
}

fn main() {
    // let input = r"example input.txt";
    let input = r"input.txt";
    let input = read_to_string(input).expect("Failed to read input").trim().to_string();
    let data: Vec<Line> = input
        .split('\n')
        .map(|x| x.trim().parse().expect("Failed to parse input"))
        .collect();

    let mut left = vec![];
    let mut right = vec![];

    for line in data {
        left.push(line.left);
        right.push(line.right);
    }

    left.sort();
    right.sort();

    let sum_of_deltas: u32 = left.iter().zip(right.iter()).map(|(a, b)| a.abs_diff(*b)).sum();

    println!("{:?}", sum_of_deltas);
}
